presto-base-arrow-flight/src/main/java/com/facebook/plugin/arrow/ArrowPageSource.java

```diff
@@ -0,0 +1,640 @@
+import org.apache.arrow.flight.Ticket;
+import org.apache.arrow.vector.BigIntVector;
+import org.apache.arrow.vector.BitVector;
+import org.apache.arrow.vector.DateDayVector;
+import org.apache.arrow.vector.DateMilliVector;
+import org.apache.arrow.vector.DecimalVector;
+import org.apache.arrow.vector.FieldVector;
+import org.apache.arrow.vector.Float4Vector;
+import org.apache.arrow.vector.Float8Vector;
+import org.apache.arrow.vector.IntVector;
+import org.apache.arrow.vector.NullVector;
+import org.apache.arrow.vector.SmallIntVector;
+import org.apache.arrow.vector.TimeMicroVector;
+import org.apache.arrow.vector.TimeMilliVector;
+import org.apache.arrow.vector.TimeSecVector;
+import org.apache.arrow.vector.TimeStampMicroVector;
+import org.apache.arrow.vector.TimeStampMilliTZVector;
+import org.apache.arrow.vector.TimeStampMilliVector;
+import org.apache.arrow.vector.TimeStampSecVector;
+import org.apache.arrow.vector.TinyIntVector;
+import org.apache.arrow.vector.VarBinaryVector;
+import org.apache.arrow.vector.VarCharVector;
+import org.apache.arrow.vector.VectorSchemaRoot;
+
+import java.math.BigDecimal;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.time.LocalTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.plugin.arrow.ArrowErrorCode.ARROW_FLIGHT_ERROR;
+
+public class ArrowPageSource
+        implements ConnectorPageSource
+{
+    private static final Logger logger = Logger.get(ArrowPageSource.class);
+    private final ArrowSplit split;
+    private final List<ArrowColumnHandle> columnHandles;
+    private boolean completed;
+    private int currentPosition;
+    private Optional<VectorSchemaRoot> vectorSchemaRoot = Optional.empty();
+    private ArrowFlightClient flightClient;
+    private FlightStream flightStream;
+
+    public ArrowPageSource(ArrowSplit split, List<ArrowColumnHandle> columnHandles, ArrowFlightClientHandler clientHandler,
+            ConnectorSession connectorSession)
+    {
+        this.columnHandles = columnHandles;
+        this.split = split;
+        getFlightStream(clientHandler, split.getTicket(), connectorSession);
+    }
+
+    private void getFlightStream(ArrowFlightClientHandler clientHandler, byte[] ticket, ConnectorSession connectorSession)
+    {
+        try {
+            Optional<String> uri = (split.getLocationUrls().isEmpty()) ?
+                    Optional.empty() : Optional.of(split.getLocationUrls().get(0));
+            flightClient = clientHandler.getClient(uri);
+            flightStream = flightClient.getFlightClient().getStream(new Ticket(ticket), clientHandler.getCallOptions(connectorSession));
+        }
+        catch (FlightRuntimeException e) {
+            throw new ArrowException(ARROW_FLIGHT_ERROR, e.getMessage(), e);
+        }
+    }
+
+    private Block buildBlockFromVector(FieldVector vector, Type type)
```

Comment:
Can we extract the conversion from Arrow to Page into a utility, and unit test that?

Reply:
Created ArrowPageUtils